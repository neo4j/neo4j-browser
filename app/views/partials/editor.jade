#editor(ng-if="show.editor", ng-controller = "EditorCtrl", ng-mousedown="focusEditor($event)", ng-class="{'file-loaded': settings.filemode && editor.document.id, 'maximize-editor': editor.maximized}")
  .file-bar.slide-down(ng-if="settings.filemode && editor.document.id")
    .title
      {{editor.document.content | autotitle}}
      //- a.btn.btn-success.btn-xs.btn-save(ng-if='editor.hasChanged()', ng-click='star()') Save
      //- a.btn.btn-default.btn-xs.btn-save(ng-if='editor.hasChanged()', ng-click='star()') Revert
    ul.file-actions.list-inline.pull-left
      li(ng-class="{'modified': editor.hasChanged()}")
        a.fa.fa-star(ng-click='star()') Save
      li
        a.fa.fa-plus(ng-click='create()') New
      li
        a.fa.fa-code-fork(ng-click='clone()') Clone
      li
        a.fa.fa-download(exportable ng-click='exportScript(editor.content)', tooltip="Export to file") Download
    ul.file-actions.list-inline.pull-right
      li
        a.button.fa.fa-play('ng-click' = 'editor.execScript(editor.content)')
      li
        a.button.fa.fa-times(ng-click='editor.setContent("")')
  .view-editor
    span(ng-class="{'one-line': editorOneLine, 'disable-highlighting': disableHighlighting}")
      .prompt.code-style $
      ui-codemirror('ui-codemirror-opts'="{gutters:['cypher-hints'], theme: 'neo', mode: 'cypher', autofocus: true, lineNumbers: true, lineWrapping: true, onLoad: codemirrorLoaded}",
        ng-model='editor.content',
        placeholder='{{motd.tip}}'
        )
    ul.controls.list-inline(ng-if="!(settings.filemode && editor.document.id)")
      li
        a.circled.favorites.sl.sl-star-circle(ng-class="{'sl-star-filled': !editor.hasChanged() && hoveringStar, 'sl-pencil-circle': editor.hasChanged() && !hoveringStar, 'sl-pencil-filled':editor.hasChanged() && hoveringStar, active: editorHasContent, 'enabled': editor.document.id, 'changed': editor.hasChanged()}", 
          ng-mouseenter="hoveringStar=true",
          ng-mouseleave="hoveringStar=false",
          ng-click='star()', 
          tooltip='Favorite', 
          tooltip-placement='left')
      li(ng-if="!settings.filemode")
        a.circled.delete.sl.sl-delete-circle(ng-click='editor.setContent("")', 
          ng-class="{active: editorHasContent, 'sl-delete-filled': hoveringDelete}", 
          ng-mouseenter="hoveringDelete=true",
          ng-mouseleave="hoveringDelete=false",
          tooltip='Clear', 
          tooltip-placement='left')
      li
        a.circled.play.sl.sl-play(ng-click = 'editor.execScript(editor.content)', 
          ng-mouseenter='hoveringPlay=true',
          ng-mouseleave='hoveringPlay=false',
          ng-class="{active: editorHasContent, 'sl-play-filled': hoveringPlay}", 
          tooltip='Play', 
          tooltip-placement='left')
    div(ng-if='!settings.onboarding')
      .message-bar.error.slide-down(ng-if='editor.showMessage', ng-class="editor.errorCode")
        span(ng-bind-html-unsafe="editor.errorMessage")
        |  Type <code click-to-code="':help commands'">:help commands</code> for a list of available commands.
        .actions
          a.fa.fa-times-circle(ng-click='editor.showMessage = null')
      .message-bar.info(ng-if='unauthorized', ng-class="editor.errorCode")
        span
          | Database access not available. Please use <code click-to-code="':server connect'">:server connect</code>
          | to establish connection. There's a graph waiting for you.
