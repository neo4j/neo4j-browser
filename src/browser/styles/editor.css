.CodeMirror-hints {
  margin: 0;
  padding: 0;
  position: absolute;
  z-index: 10;
  list-style: none;
  box-shadow: 2px 3px 5px rgba(0, 0, 0, 0.2);
  border: 1px solid silver;
  background: white;
  font-size: 90%;
  font-family: monospace;

  max-height: 30em;
  max-width: 600px;
  overflow-y: auto;
  overflow-x: auto;
}

.CodeMirror-hint {
  margin: 2px 0;
  padding: 0 4px;
  white-space: pre;
  color: #657b83;
  cursor: pointer;
  font-size: 11pt;
  background-position-x: 5px;
}

.CodeMirror-hint b {
  color: #073642;
}

.CodeMirror-hint-active {
  background-color: #efeff4;
}

.cm-hint-keyword {
}

.cm-hint-label {
  padding-left: 22px !important;
  background-size: auto 80% !important;
  background-position: 3px center;
  background-repeat: no-repeat !important;
  background-image: url("data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' version='1.1' viewBox='0 0 40 40'><circle cx='20' cy='20' r='19' stroke='black' stroke-width='1' fill='white'></circle><text x='50%' y='50%' text-anchor='middle' dy='.35em' font-size='28' font-family='Monaco' fill='%23cb4b16'>L</text></svg>");
}

.cm-hint-relationshipType {
  padding-left: 22px !important;
  background-size: auto 80% !important;
  background-position: 3px center;
  background-repeat: no-repeat !important;
  background-image: url("data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' version='1.1' viewBox='0 0 40 40'><circle cx='20' cy='20' r='19' stroke='black' stroke-width='1' fill='white'></circle><text x='50%' y='50%' text-anchor='middle' dy='.35em' font-size='28' font-family='Monaco' fill='%23cb4b16'>R</text></svg>");
}

.cm-hint-variable {
  padding-left: 22px !important;
  background-size: auto 80% !important;
  background-position: 3px center;
  background-repeat: no-repeat !important;
  background-image: url("data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' version='1.1' viewBox='0 0 40 40'><circle cx='20' cy='20' r='19' stroke='black' stroke-width='1' fill='white'></circle><text x='50%' y='50%' text-anchor='middle' dy='.35em' font-size='28' font-family='Monaco' fill='%23268bd2'>V</text></svg>");
}

.cm-hint-procedure {
  padding-left: 22px !important;
  background-size: auto 80% !important;
  background-position: 3px center;
  background-repeat: no-repeat !important;
  background-image: url("data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' version='1.1' viewBox='0 0 40 40'><circle cx='20' cy='20' r='19' stroke='black' stroke-width='1' fill='white'></circle><text x='50%' y='50%' text-anchor='middle' dy='.35em' font-size='28' font-family='Monaco' fill='%236c71c4'>λ</text></svg>");
}

.cm-hint-function {
  padding-left: 22px !important;
  background-size: auto 80% !important;
  background-position: 3px center;
  background-repeat: no-repeat !important;
  background-image: url("data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' version='1.1' viewBox='0 0 40 40'><circle cx='20' cy='20' r='19' stroke='black' stroke-width='1' fill='white'></circle><text x='50%' y='50%' text-anchor='middle' dy='.35em' font-size='28' font-family='Monaco' fill='%236c71c4'>λ</text></svg>");
}

.cm-hint-parameter {
  padding-left: 22px !important;
  background-size: auto 80% !important;
  background-position: 3px center;
  background-repeat: no-repeat !important;
  background-image: url("data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' version='1.1' viewBox='0 0 40 40'><circle cx='20' cy='20' r='19' stroke='black' stroke-width='1' fill='white'></circle><text x='50%' y='50%' text-anchor='middle' dy='.35em' font-size='28' font-family='Monaco' fill='%23dc322f'>$</text></svg>");
}

.cm-hint-propertyKey {
  padding-left: 22px !important;
  background-size: auto 80% !important;
  background-position: 3px center;
  background-repeat: no-repeat !important;
  background-image: url("data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' version='1.1' viewBox='0 0 40 40'><circle cx='20' cy='20' r='19' stroke='black' stroke-width='1' fill='white'></circle><text x='50%' y='50%' text-anchor='middle' dy='.35em' font-size='28' font-family='Monaco' fill='%23586e75'>P</text></svg>");
}

.cm-hint-consoleCommand {
  padding-left: 22px !important;
  background-size: auto 80% !important;
  background-position: 3px center;
  background-repeat: no-repeat !important;
  background-image: url("data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' version='1.1' viewBox='0 0 40 40'><circle cx='20' cy='20' r='19' stroke='black' stroke-width='1' fill='white'></circle><text x='50%' y='50%' text-anchor='middle' dy='.35em' font-size='28' font-family='Monaco' fill='%23d33682'>C</text></svg>");
}

.cm-hint-procedureOutput {
  padding-left: 22px !important;
  background-size: auto 80% !important;
  background-position: 3px center;
  background-repeat: no-repeat !important;
  background-image: url("data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' version='1.1' viewBox='0 0 40 40'><circle cx='20' cy='20' r='19' stroke='black' stroke-width='1' fill='white'></circle><text x='50%' y='50%' text-anchor='middle' dy='.35em' font-size='28' font-family='Monaco' fill='%23268bd2'>V</text></svg>");
}
