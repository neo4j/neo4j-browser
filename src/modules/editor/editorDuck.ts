export type EditorState = {
  // ... state properties
} 