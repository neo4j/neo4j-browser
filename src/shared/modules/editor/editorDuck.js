/*
 * Copyright (c) 2002-2020 "Neo4j,"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import Rx from 'rxjs/Rx'
import { getUrlParamValue } from 'services/utils'
import {
  getSettings,
  DISABLE_IMPLICIT_INIT_COMMANDS
} from 'shared/modules/settings/settingsDuck'
import { APP_START, URL_ARGUMENTS_CHANGE } from 'shared/modules/app/appDuck'
import { executeCommand } from 'shared/modules/commands/commandsDuck'

const NAME = 'editor'
export const SET_CONTENT = `${NAME}/SET_CONTENT`
export const EDIT_CONTENT = `${NAME}/EDIT_CONTENT`
export const FOCUS = `${NAME}/FOCUS`
export const EXPAND = `${NAME}/EXPAND`
export const NOT_SUPPORTED_URL_PARAM_COMMAND = `${NAME}/NOT_SUPPORTED_URL_PARAM_COMMAND`

// Supported commands
const validCommandTypes = {
  play: (cmdchar, args) => `${cmdchar}play ${args.join(' ')}`,
  edit: (_, args) => args.join('\n'),
  param: (cmdchar, args) => `${cmdchar}param ${args.join(' ')}`,
  params: (cmdchar, args) => `${cmdchar}params ${args.join(' ')}`
}

export const setContent = newContent => ({
  type: SET_CONTENT,
  message: newContent
})
export const editContent = (id, message) => ({
  type: EDIT_CONTENT,
  message,
  id
})

export const populateEditorFromUrlEpic = (some$, store) => {
  return some$
    .ofType(APP_START)
    .merge(some$.ofType(URL_ARGUMENTS_CHANGE))
    .delay(1) // Timing issue. Needs to be detached like this
    .mergeMap(action => {
      if (!action.url) {
        return Rx.Observable.never()
      }
      const cmdParam = getUrlParamValue('cmd', action.url)

      // No URL command param found
      if (!cmdParam || !cmdParam[0]) {
        return Rx.Observable.never()
      }

      // Not supported URL param command
      if (!Object.keys(validCommandTypes).includes(cmdParam[0])) {
        return Rx.Observable.of({
          type: NOT_SUPPORTED_URL_PARAM_COMMAND,
          command: cmdParam[0]
        })
      }

      const commandType = cmdParam[0]
      const cmdchar = getSettings(store.getState()).cmdchar
      // Credits to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/decodeURIComponent
      // for the "decodeURIComponent cannot be used directly to parse query parameters"
      const cmdArgs =
        getUrlParamValue(
          'arg',
          decodeURIComponent(action.url.replace(/\+/g, ' '))
        ) || []
      const fullCommand = validCommandTypes[commandType](cmdchar, cmdArgs)

      // Play command is considered safe and can run automatically
      // When running the explicit command, also set flag to skip any implicit init commands
      if (commandType === 'play') {
        return [
          executeCommand(fullCommand),
          { type: DISABLE_IMPLICIT_INIT_COMMANDS }
        ]
      }

      return Rx.Observable.of({ type: SET_CONTENT, ...setContent(fullCommand) })
    })
}
