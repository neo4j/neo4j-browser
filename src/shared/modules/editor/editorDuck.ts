/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import {
  CypherEditorSupport,
  EditorSupportCompletionItem
} from 'cypher-editor-support'
import { editor, languages } from 'monaco-editor'
import { Action } from 'redux'
import { Epic } from 'redux-observable'
import Rx from 'rxjs/Rx'

import {
  monacoDarkTheme,
  monacoLightTheme
} from 'browser/modules/Editor/CypherMonacoThemes'
import { CypherTokensProvider } from 'browser/modules/Editor/CypherTokensProvider'
import cypherFunctions from 'browser/modules/Editor/cypher/functions'
import {
  toFunction,
  toLabel,
  toProcedure,
  toPropertyKey,
  toRelationshipType
} from 'browser/modules/Editor/editorSchemaConverter'
import consoleCommands from 'browser/modules/Editor/language/consoleCommands'
import { getUrlParamValue } from 'services/utils'
import { GlobalState } from 'shared/globalState'
import { APP_START, URL_ARGUMENTS_CHANGE } from 'shared/modules/app/appDuck'
import {
  commandSources,
  executeCommand
} from 'shared/modules/commands/commandsDuck'
import {
  DARK_THEME,
  DISABLE_IMPLICIT_INIT_COMMANDS,
  LIGHT_THEME,
  OUTLINE_THEME
} from 'shared/modules/settings/settingsDuck'

export const SET_CONTENT = 'editor/SET_CONTENT'
export const EDIT_CONTENT = 'editor/EDIT_CONTENT'
export const FOCUS = 'editor/FOCUS'
export const EXPAND = 'editor/EXPAND'
export const NOT_SUPPORTED_URL_PARAM_COMMAND =
  'editor/NOT_SUPPORTED_URL_PARAM_COMMAND'

// Supported commands
const validCommandTypes: { [key: string]: (args: string[]) => string } = {
  play: args => `:play ${args.join(' ')}`,
  guide: args => `:guide ${args.join(' ')}`,
  edit: args => args.join('\n'),
  param: args => `:param ${args.join(' ')}`,
  params: args => `:params ${args.join(' ')}`
}

interface SetContentAction {
  type: typeof SET_CONTENT
  message: string
}

export const setContent = (message: string): SetContentAction => ({
  type: SET_CONTENT,
  message
})

interface EditContentAction {
  type: typeof EDIT_CONTENT
  message: string
  id: string
  isProjectFile: boolean
  isStatic: boolean
  name: null | string
  directory: null
}

interface EditContentOptions {
  name?: string | null
  isStatic?: boolean
  isProjectFile?: boolean
}

export const editContent = (
  id = '',
  message: string,
  {
    name = null,
    isStatic = false,
    isProjectFile = false
  }: EditContentOptions = {}
): EditContentAction => ({
  type: EDIT_CONTENT,
  message,
  id,
  isProjectFile,
  isStatic,
  name,
  directory: null
})

export const populateEditorFromUrlEpic: Epic<any, GlobalState> = some$ => {
  return some$
    .ofType(APP_START)
    .merge(some$.ofType(URL_ARGUMENTS_CHANGE))
    .delay(1) // Timing issue. Needs to be detached like this
    .mergeMap(action => {
      if (!action.url) {
        return Rx.Observable.never()
      }
      const cmdParam = getUrlParamValue('cmd', action.url)

      // No URL command param found
      if (!cmdParam || !cmdParam[0]) {
        return Rx.Observable.never()
      }

      // Not supported URL param command
      if (!Object.keys(validCommandTypes).includes(cmdParam[0])) {
        return Rx.Observable.of({
          type: NOT_SUPPORTED_URL_PARAM_COMMAND,
          command: cmdParam[0]
        })
      }

      const commandType = cmdParam[0]
      // Credits to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/decodeURIComponent
      // for the "decodeURIComponent cannot be used directly to parse query parameters"
      const cmdArgs =
        getUrlParamValue(
          'arg',
          decodeURIComponent(action.url.replace(/\+/g, ' '))
        ) || []
      const fullCommand = validCommandTypes[commandType](cmdArgs)

      // Play command is considered safe and can run automatically
      // When running the explicit command, also set flag to skip any implicit init commands

      if (['play', 'guide'].includes(commandType)) {
        return [
          executeCommand(fullCommand, { source: commandSources.url }),
          { type: DISABLE_IMPLICIT_INIT_COMMANDS }
        ]
      }

      return Rx.Observable.of(setContent(fullCommand))
    })
}

// CypherEditorSupport returns the content attributes of procedures with dots wrapped in backticks, e.g. "`apoc.coll.avg`"
// This function strips any surrounding backticks before we use the .content value in the completion item provider
const stripSurroundingBackticks = (str: string) =>
  str.charAt(0) === '`' && str.charAt(str.length - 1) === '`'
    ? str.substr(1, str.length - 2)
    : str

export const getText = (item: EditorSupportCompletionItem): string =>
  ['function', 'procedure'].includes(item.type)
    ? stripSurroundingBackticks(item.content)
    : item.content

const editorSupport = new CypherEditorSupport('')

export const initializeCypherEditorEpic: Epic<
  { type: typeof APP_START },
  GlobalState
> = actions$ => {
  return actions$
    .ofType(APP_START)
    .take(1)
    .do(() => {
      languages.register({ id: 'cypher' })
      languages.setTokensProvider('cypher', new CypherTokensProvider())
      languages.setLanguageConfiguration('cypher', {
        brackets: [
          ['(', ')'],
          ['{', '}'],
          ['[', ']'],
          ["'", "'"],
          ['"', '"']
        ],
        comments: {
          blockComment: ['/*', '*/'],
          lineComment: '//'
        }
      })

      languages.registerCompletionItemProvider('cypher', {
        triggerCharacters: ['.', ':', '[', '(', '{', '$'],
        provideCompletionItems: (model, position) => {
          const { startColumn, endColumn } =
            model.getWordUntilPosition(position)
          const range = {
            startLineNumber: position.lineNumber,
            endLineNumber: position.lineNumber,
            startColumn,
            endColumn
          }
          editorSupport.update(model.getValue())
          let items: EditorSupportCompletionItem[] = []
          // Cypher editor support repo has internal type errors
          try {
            items = editorSupport.getCompletion(
              position.lineNumber,
              position.column - 1
            ).items
          } catch {}

          const { CompletionItemKind } = languages
          const completionTypes: Record<string, languages.CompletionItemKind> =
            {
              keyword: CompletionItemKind.Keyword,
              label: CompletionItemKind.Field,
              relationshipType: CompletionItemKind.Reference,
              variable: CompletionItemKind.Variable,
              procedure: CompletionItemKind.Function,
              function: CompletionItemKind.Function,
              parameter: CompletionItemKind.TypeParameter,
              propertyKey: CompletionItemKind.Property,
              consoleCommand: CompletionItemKind.Function,
              consoleCommandSubcommand: CompletionItemKind.Function,
              procedureOutput: CompletionItemKind.Operator
            }

          // word preceding trigger character, used to determine range (where to insert) procedure suggestions
          const { word } = model.getWordUntilPosition({
            lineNumber: position.lineNumber,
            column: position.column - 1
          })

          const getRange = (type: string, text: string) =>
            ['consoleCommand', 'label', 'relationshipType'].includes(type)
              ? { ...range, startColumn: range.startColumn - 1 }
              : ['function', 'procedure'].includes(type)
              ? {
                  ...range,
                  startColumn:
                    range.startColumn -
                    (text.lastIndexOf(word) + word.length + 1)
                }
              : range

          return {
            suggestions: items.map((item, index) => {
              const label = getText(item)
              return {
                label,
                kind: completionTypes[item.type],
                insertText: label,
                range: getRange(item.type, label),
                detail: item.postfix || undefined,
                sortText: encodeNumberAsSortableString(index)
              }
            })
          }
        }
      })

      editor.defineTheme(DARK_THEME, monacoDarkTheme)
      editor.defineTheme(LIGHT_THEME, monacoLightTheme)
      editor.defineTheme(OUTLINE_THEME, monacoLightTheme)
      // Browser's light theme is called 'normal', but OS's light theme is called 'light'
      // 'light' is used when theme is set to light in OS and auto in browser
      editor.defineTheme('light', monacoLightTheme)
    })
    .ignoreElements()
}
function encodeNumberAsSortableString(number: number): string {
  // use letter prefix to encode numbers. breaks after numbers are have more than ~28 digits
  // Monaco by default sorts its suggestion by label text. But our language support sorts
  // after relevance first and then alphabetically.
  const A_CHAR = 65
  const prefix = String.fromCharCode(A_CHAR + number.toString().length)
  return `${prefix}${number}`
}

export const updateEditorSupportSchemaEpic: Epic<Action, GlobalState> = (
  actions$,
  store
) =>
  actions$
    .do(() => {
      const state = store.getState()
      const schema = {
        consoleCommands,
        parameters: Object.keys(state.params),
        labels: state.meta.labels.map(toLabel),
        relationshipTypes: state.meta.relationshipTypes.map(toRelationshipType),
        propertyKeys: state.meta.properties.map(toPropertyKey),
        functions: [
          ...cypherFunctions,
          ...state.meta.functions.map(toFunction)
        ],
        procedures: state.meta.procedures.map(toProcedure)
      }
      editorSupport.setSchema(schema)
    })
    .ignoreElements()
