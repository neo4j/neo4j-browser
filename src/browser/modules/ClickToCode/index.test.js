/*
 * Copyright (c) 2002-2019 "Neo4j,"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import { render, fireEvent } from '@testing-library/react'

import { createBus } from 'suber'

import { ClickToCode } from './index'
import { SET_CONTENT } from 'shared/modules/editor/editorDuck'

describe('ClickToCode', () => {
  let bus
  beforeEach(() => {
    bus = createBus()
  })
  afterEach(() => {
    bus.reset()
  })
  test('does not render if no children', () => {
    // Given
    const myFn = jest.fn()
    const MyComp = () => <span>yo!</span>

    // When
    bus.take(SET_CONTENT, myFn)
    const { container } = render(
      <ClickToCode CodeComponent={MyComp} bus={bus} />
    )

    // Then
    expect(container).toMatchSnapshot()
    expect(myFn).toHaveBeenCalledTimes(0)
  })
  test('renders if no CodeComponent is provided', () => {
    // Given
    const myFn = jest.fn()

    // When
    bus.take(SET_CONTENT, myFn)
    const { container } = render(<ClickToCode bus={bus} />)

    // Then
    expect(container).toMatchSnapshot()
    expect(myFn).toHaveBeenCalledTimes(0)
  })
  test('renders code as the code when code is available', () => {
    // Given
    const myFn = jest.fn()
    const code = 'my code'

    // When
    bus.take(SET_CONTENT, myFn)
    const { container, getByText } = render(
      <ClickToCode code={code} bus={bus}>
        hello
      </ClickToCode>
    )

    // Then
    expect(container).toMatchSnapshot()
    expect(myFn).toHaveBeenCalledTimes(0)

    // When
    fireEvent.click(getByText('hello'))

    // Then
    expect(myFn).toHaveBeenCalledTimes(1)
    expect(myFn).toHaveBeenCalledWith({ message: code, type: SET_CONTENT })
  })
  test('renders children as code if no code is provided', () => {
    // Given
    const myFn = jest.fn()
    const childrenString = 'hellohi!'

    // When
    bus.take(SET_CONTENT, myFn)
    const { container, getByText } = render(
      <ClickToCode bus={bus}>hellohi!</ClickToCode>
    )

    // Then
    expect(container).toMatchSnapshot()
    expect(myFn).toHaveBeenCalledTimes(0)

    // When
    fireEvent.click(getByText(childrenString))

    // Then
    expect(myFn).toHaveBeenCalledTimes(1)
    expect(myFn).toHaveBeenCalledWith({
      message: childrenString,
      type: SET_CONTENT
    })
  })
  test('renders all children', () => {
    // Given
    const code = 'my code'
    const myFn = jest.fn()
    const children = (
      <div>
        <span>hello</span>
        hi!
      </div>
    )

    // When
    bus.take(SET_CONTENT, myFn)
    const { container, getByText } = render(
      <ClickToCode code={code} bus={bus}>
        {children}
      </ClickToCode>
    )

    // Then
    expect(container).toMatchSnapshot()
    expect(myFn).toHaveBeenCalledTimes(0)

    // When
    fireEvent.click(getByText('hello'))

    // Then
    expect(myFn).toHaveBeenCalledTimes(1)
    expect(myFn).toHaveBeenCalledWith({
      message: code,
      type: SET_CONTENT
    })
  })
})
