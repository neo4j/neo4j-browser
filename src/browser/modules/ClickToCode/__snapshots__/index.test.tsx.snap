// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClickToCode does not render if no children 1`] = `<div />`;

exports[`ClickToCode renders all children 1`] = `
<div>
  <code
    class="sc-bdVaJa biJQaZ"
  >
    <div>
      <span>
        hello
      </span>
      hi!
    </div>
  </code>
</div>
`;

exports[`ClickToCode renders children as code if no code is provided 1`] = `
<div>
  <code
    class="sc-bdVaJa biJQaZ"
  >
    hellohi!
  </code>
</div>
`;

exports[`ClickToCode renders code as the code when code is available 1`] = `
<div>
  <code
    class="sc-bdVaJa biJQaZ"
  >
    hello
  </code>
</div>
`;

exports[`ClickToCode renders if no CodeComponent is provided 1`] = `<div />`;
